zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java

```diff
@@ -279,22 +284,36 @@ public void setDefaultInterpreterGroup(String defaultInterpreterGroup) {
     this.defaultInterpreterGroup = defaultInterpreterGroup;
   }
 
-  public Map<String, Object> getNoteParams() {
-    return noteParams;
+  public ConcurrentHashMap<String, Object> getNoteParams() {
+    if (noteParams == null) {
+      return new ConcurrentHashMap<>();
+    }
+    return new ConcurrentHashMap<>(noteParams);
```

Comment:
Why return a new ConcurrentHashMap?
I don't know if returning an Collections.unmodifiableMap might be a good approach here.
This [code part ](https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java#L450-L455) should be the only place where the map is edited from another class. Here qw could force the class to create a new map to work on and then communicate it to the Note.